### PR TITLE
Test display

### PR DESCRIPTION
**Pull request by albert4750**
Display first implementation done

TODO:
switch GPIO and SPI used by library
currently added ignore for the two library files but we should review whether or not thats ok

### CODE REVIEW

File: revolution/environment.py [hunk 1]
```diff
@@ -71,6 +71,48 @@ class Context:
...
+    ppt_status: bool = True
+    rad_status: bool = True
+    # less important display flags
+    eco_mode: bool = False  # eco mode that draws leaf symbol on d1
+    direction: bool = True  # direction of the car, True is forward, else rev
+    headlight: bool = False  # whether headlights are on
+    # might have replacement but for display
+    hazard_lights_status: bool = False
+    left_indicator_light_status: bool = False
+    right_indicator_light_status: bool = False
```
Comment (AussieSeaweed) [inline revolution/environment.py:115]: Some of these attributes are already present.

File: revolution/st7565.py [hunk 2]
```diff
@@ -0,0 +1,1057 @@
+# flake8: noqa
+# type: ignore
+from __future__ import print_function
```
Comment (AussieSeaweed) [inline revolution/st7565.py:3]: Not necessary

File: revolution/st7565.py [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+        """
+        self.send_command([self.CMD_SET_VOLUME_FIRST])
+        self.send_command([self.CMD_SET_VOLUME_SECOND | (level & 0x3f)])
+
+    def clear_back_buffer(self):
+        """Clear back buffer only"""
+        self.back_buffer = np.zeros((self.LCD_HEIGHT, self.LCD_WIDTH), dtype='uint8')
+
+    def init(self):
+        import RPi.GPIO as GPIO
```
Comment (AussieSeaweed) [inline revolution/st7565.py:340]: We can't just import RPi like this. We are not using Raspberry Pi

File: revolution/xglcd_font.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        """
+         
+        self.width = width
+        self.height = height
+        self.start_letter = start_letter
+        self.letters = self.__load_xglcd_font(path)
+        self.height_bytes = int((self.letters.shape[1] - 1) / width)
+
+                      
+    def __load_xglcd_font(self, path):
```
Comment (AussieSeaweed) [inline revolution/xglcd_font.py:39]: This is crazy lol

File: revolution/display.py [hunk 5]
```diff
@@ -1,15 +1,379 @@
...
+        self.d1.init()
+        self.d2.init()
+        # TODO generate new JetBrains13x21 font to have all characters
+        self.JetBrains13x20 = font.XglcdFont("display_data/JetBrains_Mono13x20.c", 13, 20, 97)  # noqa: E501
+        self.JetBrains13x21 = font.XglcdFont("display_data/JetBrains_Mono13x21_Symbol.c", 13, 21, 32)  # noqa: E501
+        self.Liberation_Sans20x28 = font.XglcdFont("revolution/display_data/Liberation_Sans20x28_Numbers.c", 20, 28, 46)  # noqa: E501
+        self.Font5x7 = font.XglcdFont("display_data/font5x7.c", 5, 7, 32)
+
+    # draw blue sky logo, wait for a while, then draw the default screen
+    def startup(self) -> None:
```
Comment (AussieSeaweed) [inline revolution/display.py:43]: Should be run as a separate thread. This was done in def setup in other ``Application``s.

**Review by AussieSeaweed — CHANGES_REQUESTED** [review]
The logic is there, but this won't run as it stands. We can discuss tomorrow.

File: revolution/environment.py [hunk 6]
```diff
@@ -71,6 +71,48 @@ class Context:
...
+    solar_current: float = 0
+    motor_power: float = 0
+    motor_voltage: float = 0
+    motor_current: float = 0
+    battery_power: float = 0
+    battery_voltage: float = 0
+    battery_current: float = 0
+    battery_soc: float = 0
+    speed_kph: float = 0
+    motor_state: int = 0  # 0=OFF, 1=PEDAL, 2=CRUISE, 3=REGEN
```
Comment (TristanRobitaille) [inline revolution/environment.py:86]: Enum for this and the ones below would be more readable

**Pull request closed without merging**